cirq/ops/linear_combinations.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, Mapping, Optional, Tuple
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+
+class LinearCombinationOfGates(value.LinearDict[raw_types.Gate]):
+    """Represents linear operator defined by a linear combination of gates.
+
+    Note that the operator may not be unitary or even normal. Examples of
+    operators that may be represented as LinearCombinationOfGates:
+     * Hamiltonians,
+     * measurement operators,
+     * Kraus operators,
+     * creation and annihilation operators.
+    """
+    def __init__(self, terms: Mapping[raw_types.Gate, value.Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of gates to coefficients in the linear combination
+                being initialized.
+        """
+        super().__init__(terms, validator=self._is_compatible)
```

Comment:
I'm always surprised that this works in python. In C#, referring to a method on `self` before the parent class has been initialized is a compile time error.

Reply:
Well, in python even typos in method names aren't errors until you try to execute them. Compared to that full initialization of the target object seems like a minor concern.